java/ql/src/experimental/Security/CWE/CWE-348/UseOfLessTrustedSource.java

```diff
@@ -0,0 +1,80 @@
+import javax.servlet.http.HttpServletRequest;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+@Controller
+public class UseOfLessTrustedSource {
+
+    private static final Logger log = LoggerFactory.getLogger(UseOfLessTrustedSource.class);
+
+    @Autowired
+    private HttpServletRequest request;
+
+    @GetMapping(value = "bad1")
+    public void bad1(HttpServletRequest request) {
+        String ip = request.getHeader("X-Forwarded-For");
+
+        log.debug("getClientIP header X-Forwarded-For:{}", ip);
+
+        if (StringUtils.isBlank(ip) || StringUtils.equalsIgnoreCase("unknown", ip)) {
+            ip = request.getHeader("Proxy-Client-IP");
+            log.debug("getClientIP header Proxy-Client-IP:{}", ip);
+        }
+        if (StringUtils.isBlank(ip) || StringUtils.equalsIgnoreCase("unknown", ip)) {
+            ip = request.getHeader("WL-Proxy-Client-IP");
+            log.debug("getClientIP header WL-Proxy-Client-IP:{}", ip);
+        }
+        if (StringUtils.isBlank(ip) || StringUtils.equalsIgnoreCase("unknown", ip)) {
+            ip = request.getHeader("HTTP_CLIENT_IP");
+            log.debug("getClientIP header HTTP_CLIENT_IP:{}", ip);
+        }
+        if (StringUtils.isBlank(ip) || StringUtils.equalsIgnoreCase("unknown", ip)) {
+            ip = request.getHeader("HTTP_X_FORWARDED_FOR");
+            log.debug("getClientIP header HTTP_X_FORWARDED_FOR:{}", ip);
+        }
+        if (StringUtils.isBlank(ip) || StringUtils.equalsIgnoreCase("unknown", ip)) {
+            ip = request.getHeader("X-Real-IP");
+            log.debug("getClientIP header X-Real-IP:{}", ip);
+        }
+        if (StringUtils.isBlank(ip) || StringUtils.equalsIgnoreCase("unknown", ip)) {
+            ip = request.getRemoteAddr();
+            log.debug("getRemoteAddr IP:{}", ip);
+        }
```

Comment:
IMHO Logging is a bad example of this vulnerability. `bad2` is a better case because it shows logic-flow stopping logic being gated behind this value. Purely logging like this doesn't seem like the best demonstration of this for the CodeQL documentation.

Reply:
There will not be a direct security issue here. However, it is very bad to do a log investigation in the online environment. For example, when the online environment is attacked, when the log is checked, the ip is forged, so the attacker cannot be found ip or increase the investigation time.